src/python_testing/TC_DEM_2_4.py

```diff
@@ -0,0 +1,367 @@
+        self.step("12")
+        await self.send_resume_request_command()
+        event_data = events_callback.wait_for_event_report(Clusters.DeviceEnergyManagement.Events.Resumed)
+        asserts.assert_equal(event_data.cause, Clusters.DeviceEnergyManagement.Enums.CauseEnum.kCancelled)
+
+        self.step("12a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("12b")
+        forecast = await self.read_dem_attribute_expect_success(attribute="Forecast")
+        asserts.assert_equal(forecast.isPausable, True)
+        asserts.assert_greater(forecast.slots[0].minPauseDuration, 1)
+        asserts.assert_greater(forecast.slots[0].maxPauseDuration, 1)
+        asserts.assert_equal(forecast.slots[0].slotIsPausable, True)
+        asserts.assert_equal(forecast.slots[1].slotIsPausable, False)
+        asserts.assert_equal(forecast.activeSlotNumber, 0)
+        asserts.assert_equal(forecast.forecastUpdateReason,
+                             Clusters.DeviceEnergyManagement.Enums.ForecastUpdateReasonEnum.kLocalOptimization)
+
+        self.step("13")
+        await self.send_pause_request_command(forecast.slots[0].minPauseDuration,
+                                              Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kLocalOptimization)
+        event_data = events_callback.wait_for_event_report(Clusters.DeviceEnergyManagement.Events.Paused)
+
+        self.step("13a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kPaused)
+
+        self.step("13b")
+        forecast = await self.read_dem_attribute_expect_success(attribute="Forecast")
+        asserts.assert_equal(forecast.forecastUpdateReason,
+                             Clusters.DeviceEnergyManagement.Enums.ForecastUpdateReasonEnum.kLocalOptimization)
+
+        self.step("14")
+        await self.send_resume_request_command()
+        event_data = events_callback.wait_for_event_report(Clusters.DeviceEnergyManagement.Events.Resumed)
+        asserts.assert_equal(event_data.cause, Clusters.DeviceEnergyManagement.Enums.CauseEnum.kCancelled)
+
+        self.step("14a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("15")
+        await self.send_pause_request_command(forecast.slots[0].minPauseDuration,
+                                              Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kLocalOptimization)
+        event_data = events_callback.wait_for_event_report(Clusters.DeviceEnergyManagement.Events.Paused)
+
+        self.step("15a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kPaused)
+
+        self.step("16")
+        logging.info(f"Sleeping for forecast.slots[0].minPauseDuration {forecast.slots[0].minPauseDuration}s")
+        time.sleep(forecast.slots[0].minPauseDuration)
```

Comment:
This isn't waiting for the Resumed Event - it could do this with a blocking timeout in the expect event instead of a sleep.

Reply:
Done